text-autospace: implement ideograph-alpha for complex path within an text run

vitorroriz opened the pull request:
#### 44e62df329a07ba50ca5feaad232f284705f4a63
<pre>
text-autospace: implement ideograph-alpha for complex path within an text run
<a href="https://bugs.webkit.org/show_bug.cgi?id=277716">https://bugs.webkit.org/show_bug.cgi?id=277716</a>
<a href="https://rdar.apple.com/133309470">rdar://133309470</a>

Reviewed by Matthieu Dubet.

This patch implements the processing of text-autospace: ideogram-alpha
only within an element. We don&apos;t yet handle element boundaries here.

Although we pass SpacingState context from one ComplexTextController
to another, we do that here in a limited way, only for measuring text for
layout and for painting. There are other places in code which this will
be necessary, for example, for handling element boundaries.

1. During the construction of ComplexTextController, we call ::adjustGlyphsAndAdvances
which already iterates through glyphs and adjust spacing for other reasons.
Now we process each pair of characters related to these glyphs here, adding the
spacing necessary before the &quot;current&quot; character.  For that reason, the SpacingState
stores information about the previous character of a run. We also save the measured
spacing in a new parallel vector m_textAutoSpaceSpacings. At this phase we can
only manipulate a glyph advance, however, for adding space &quot;before&quot; a glyph,
we need to move the glyph to the logical right, which is done later on ::advance.

2. ComplexTextController::advance is called for both layout and painting, but during
painting it has access to a GlyphBuffer and it add glyphs into it. We are introducing
a new GlyphBuffer::add function that also takes the glyph&apos;s origin, so we can manipulate
the origin as necessary by adding the previous calculated spacing.

3. Doing #1 and #2 is already enough for painting the extra spacing between relevant characters
according to their classes. Howeverm the width measured during layout would be broken because
IFC splits text content into inlineTextItem(s) and measure the width of each item independently.
This means that we already have to handle SpacingState passing here, otherwise we are not able
to handle spacing between characters on the boundary of different InlineTextItem.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001.html: Added.
* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isPunctuation):
(WTF::isOpeningPunctuation):
(WTF::isClosingPunctuation):
(WTF::isOfScriptType):
(WTF::eastAsianWidth):
(WTF::isEastAsianFullWidth):
(WTF::isCJKSymbolOrPunctuation):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::width):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.h:
(WebCore::Layout::TextUtil::width):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::ComplexTextController):
(WebCore::ComplexTextController::advance):
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/ComplexTextController.h:
(WebCore::ComplexTextController::ComplexTextRun::textAutospaceSize const):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::width const):
(WebCore::FontCascade::codePath const):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::add):
* Source/WebCore/platform/graphics/TextRun.cpp:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::add):
(WebCore::WidthCache::invalidateCacheForTextSpacing):
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/text/TextSpacing.cpp: Added.
(WebCore::TextAutospace::shouldApplySpacing const):
(WebCore::TextAutospace::textAutospaceSize):
(WebCore::TextSpacing::isIdeograph):
(WebCore::TextSpacing::isNonIdeographicNumeral):
(WebCore::TextSpacing::characterClass):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextAutospace::hasIdeographAlpha const):
(WebCore::TextAutospace::hasIdeographNumeric const):

Canonical link: <a href="https://commits.webkit.org/282192@main">https://commits.webkit.org/282192@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25358971ede027a54a2545da6c50c3512f44948d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66381 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12948 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13282 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8940 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54054 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/35508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11362 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11877 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/55494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57157 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68110 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61640 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6343 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54074 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/57867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13865 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5284 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83403 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37552 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14642 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/39734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,110 @@
+
+using namespace TextSpacing;
+
+unsigned TextAutospace::numberOfAdditionalSpacingUnits(char32_t firstCharacter, char32_t secondCharacter) const
```
If this only ever returns 0 or 1, why is it `numberofAdditionalSpacingUnits`?

fantasai commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -24,10 +24,34 @@
+// Classes are defined at https://www.w3.org/TR/css-text-4/#text-spacing-classes
+CharacterClass characterClass(char32_t character);
+struct SpacingState {
+    char32_t lastCharacterFromPreviousRun { 0 };
```
Should this also maintain its classification state?

vitorroriz replied:
We could, but most of our state passing will happen on higher level code (not in ComplexTextController), so we would have to classify the code point anyway, so I guess it is not beneficial.

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,110 @@
+    auto secondCharacterClass = characterClass(secondCharacter);
+
+    if (hasIdeographAlpha() && ((firstCharacterClass == CharacterClass::Ideograph && secondCharacterClass == CharacterClass::NonIdeographLetter) || (firstCharacterClass == CharacterClass::NonIdeographLetter && secondCharacterClass == CharacterClass::Ideograph)))
+        ++result;
```
I think this boolean is very long and should be line-wrapped. :)

To optimize this check, you can try using some of the techniques from `BreakLines::nextBreakablePosition()`'s "Non-ASCII rapid lookup."

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,110 @@
+
+static bool isIdeograph(char32_t character)
+{
+    // All characters in the range of U+3041 to U+30FF, except those that belong to Unicode Punctuation [P*]
+    if ((character >= 0x3041 && character <= 0x30FF) && !isPunctuation(character))
+        return true;
+    // CJK Strokes (U+31C0 to U+31EF).
+    if (character >= 0x31C0 && character <= 0x31EF)
+        return true;
+    // Katakana Phonetic Extensions (U+31F0 to U+31FF)
+    if (character >= 0x31F0 && character <= 0x31FF)
+        return true;
+    if (isOfScriptType(character, UScriptCode::USCRIPT_HAN))
+        return true;
```
I think you can include almost every character in the range `0x2E80 <= character && character <= 0xA4CF` here, excepting punctuation etc. which you can then filter out with the general category mask.

It's better to not have a chain of conditionals, because it's slow, so to avoid slowing down non-CJK text, have a higher-level conditional to capture the range, and do more analysis within it. (You can also use bit hacks to filter a particular block, if some of the checks within it are more expensive.)

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -748,6 +751,26 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+            if (textAutoSpace.hasIdeographAlpha()) {
+                char32_t previousCharacter = 0;
+                if (glyphIndex) {
+                    auto previousCharacterIndex = complexTextRun.indexAt(glyphIndex - 1);
```
For the purposes of text-spacing, the character we care about is the base character. So you need to skip over combining marks.

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,110 @@
+
+    // FIXME: implement remaining classes for text-autospace: punctuation
+    return CharacterClass::Other;
+}
```
The if-clauses in this function should be structured more efficiently.

- Since you have a category mask, use it more aggressively to filter things out. For example, you can check for [U_GC_P_MASK](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uchar_8h.html#a84973d3e336d9858aac0bc31d39ffc5f) before doing more specific punctuation checks.
- Combine bitmasks together so you can check them simultaneously, e.g. (U_GC_M_MASK | U_GC_L_MASK)
- Exit early for things you know are not relevant, if you're already incorporating a conditional check for them. For example, once have checked for (U_GC_M_MASK | U_GC_L_MASK) we know that none of the rest of the function applies, so we can return instead of falling through to the punctuation checks. Because of short-circuiting, every clause in || or && is essentially an if statement -- if we know what to do for the else case, make that if-else explicit and make the short-circuit short-circuit the rest of the function.

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -748,6 +751,26 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+            const auto& textAutoSpace =  m_font.textAutospace();
+            float textAutoSpaceSpacing = 0;
+            if (textAutoSpace.hasIdeographAlpha()) {
+                char32_t previousCharacter = 0;
```
Shouldn't this be `UChar`?

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -748,6 +751,26 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+                    auto previousCharacterIndex = complexTextRun.indexAt(glyphIndex - 1);
+                    previousCharacter = static_cast<char32_t>(*(charactersPointer + previousCharacterIndex));
+                } else
+                    previousCharacter = m_textSpacingState.lastCharacterFromPreviousRun;
```
We are in a loop here. Wouldn't it make more sense to just keep track of the previous character? Initialize it to `previousCharacter = m_textSpacingState.lastCharacterFromPreviousRun` before the start of the loop, and then update it as we go.

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -835,4 +858,12 @@ ComplexTextController::ComplexTextRun::ComplexTextRun(const Vector<FloatSize>& a
+        m_measuredTextSpacing =  0.125 * font.fontMetrics().ideogramWidth().value_or(0);
+
+    return m_measuredTextSpacing.value();
+}
```
If the font is something that changes over the course of the run, then this is going to return the wrong results. And if it's not, it makes more sense to just compute it up front and leave it there rather than doing this kind of optional caching, no?

vitorroriz replied:
The font shouldn't change within a `ComplexTextRun`, but you are right, we should compute it up front on the creation of `ComplexTextRun`.

fantasai reviewed (commented):
Left a bunch of comments about optimizing the character class processing, but feel free to address that in follow-up PRs.

fantasai commented on Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
```diff
@@ -45,6 +45,7 @@
     , m_glyphCount(CTRunGetGlyphCount(ctRun))
     , m_stringLocation(stringLocation)
     , m_isLTR(!(CTRunGetStatus(ctRun) & kCTRunStatusRightToLeft))
+    , m_measuredTextSpacing(0.125 * font.fontMetrics().ideogramWidth().value_or(0))
```
Should this be `m_textAutospaceSize` or something?

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -777,6 +791,7 @@ void ComplexTextController::adjustGlyphsAndAdvances()
             glyphOrigin.move(advance);
 
             lastCharacterIndex = characterIndex;
+            previousCharacter = character;
```
If these two are related, should they have related names?

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -633,6 +636,7 @@ void ComplexTextController::adjustGlyphsAndAdvances()
     bool runForbidsLeftExpansion = m_run.expansionBehavior().left == ExpansionBehavior::Behavior::Forbid;
     bool runForbidsRightExpansion = m_run.expansionBehavior().right == ExpansionBehavior::Behavior::Forbid;
 
+    UChar previousCharacter = m_textSpacingState.lastCharacterFromPreviousRun;
```
Good, but you should also pull out the character's class and cache it, otherwise we're calculating that twice per character and we really only need to do it once.

fantasai commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -629,6 +629,10 @@ FontCascade::CodePath FontCascade::codePath(const TextRun& run, std::optional<un
     if (s_codePath != CodePath::Auto)
         return s_codePath;
 
+    // FIXME: add support for text-autospace on simple path (rdar://133319627).
+    if (textAutospace().hasIdeographAlpha())
+        return CodePath::Complex;
+
```
One thing we could do is, while we're scanning the text for a complex character, also scan it for ideographs -- if there are none, then we can continue on the simple path.

vitorroriz replied:
Good idea but I made it simple here because I don't intend to let it just on the Complex path for long.

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,111 @@
+        if (firstCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::NonIdeographLetter;
+        if (firstCharacterClass == CharacterClass::NonIdeographLetter && secondCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::Ideograph;
```
If you make the character classes each a separate bit (see BreakLines::classify & co), then you can reduce this to a single conditional check: `return (firstCharacterClass | secondCharacterClass == CharacterClass::NonIdeographLetter | CharacterClass::Ideograph)`.

fantasai commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,111 @@
+        return CharacterClass::Other;
+    }
+
+    if (isNonIdeographicNumeral(character, generalCategoryMask))
+        return CharacterClass::NonIdeographNumeral;
```
We already know it's not ideographic, right? So shouldn't this just check if it's a numeral?

vitorroriz replied:
I'm not sure...Spec says:

> non-ideographic numerals
> Includes all [typographic character](https://www.w3.org/TR/css-text-4/#typographic-character-unit) units that belong to the Unicode Decimal Digit Number [Nd] [general category](https://www.w3.org/TR/css-text-4/#unicode-general-category), except when any of the following conditions are met:
> is categorized as East Asian Fullwidth (F) by [[UAX11]](https://www.w3.org/TR/css-text-4/#biblio-uax11).
> is upright in vertical text flow using the [text-orientation](https://www.w3.org/TR/css-writing-modes-4/#propdef-text-orientation) property or the [text-combine-upright](https://www.w3.org/TR/css-writing-modes-4/#propdef-text-combine-upright) property.

So I guess we can't do that unless we can guarantee that a character that is not a ideograph can't be classified as East Asian fullwidth.

fantasai replied:
Ah, right.

mdubet commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,105 @@
+    if (hasIdeographAlpha())
+        return (static_cast<uint8_t>(firstCharacterClass) | static_cast<uint8_t>(secondCharacterClass)) == ideographAlphaMask;
+    return false;
+}
```
Missing a newline here

mdubet commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,105 @@
+static bool isNonIdeographicNumeral(char32_t character, uint32_t generalCategoryMask)
+{
+    // FIXME: Should also check that it is not: upright in vertical text flow using the text-orientation property or the text-combine-upright property.
+    return ((generalCategoryMask & U_GC_ND_MASK) && !isEastAsianFullWidth(character));
```
Extra parenthesis

mdubet commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -793,6 +809,7 @@ ComplexTextController::ComplexTextRun::ComplexTextRun(const Font& font, const UC
     , m_indexEnd(indexEnd)
     , m_stringLocation(stringLocation)
     , m_isLTR(ltr)
+    , m_textAutospaceSize(0.125 * font.fontMetrics().ideogramWidth().value_or(0))
```
What is this magic number `0.125` ? We should define a name for this magic constant.

mdubet commented on Source/WebCore/platform/graphics/ComplexTextController.h:
```diff
@@ -169,6 +172,7 @@ class ComplexTextController {
     Vector<FloatSize, 256> m_adjustedBaseAdvances;
     Vector<FloatPoint, 256> m_glyphOrigins;
     Vector<CGGlyph, 256> m_adjustedGlyphs;
+    Vector<float, 256> m_textAutoSpaceSpacings;
```
Why `256` ?

vitorroriz replied:
All our parallel vectors at `ComplexTextRun` are initialized with a 256 capacity, this will have the same size as the others as this is another parallel vector. We might consider recalculating what is the optimal initial capacity but not here.

```
    Vector<FloatSize, 256> m_adjustedBaseAdvances;
    Vector<FloatPoint, 256> m_glyphOrigins;
    Vector<CGGlyph, 256> m_adjustedGlyphs;
```

mdubet replied:
This could use a follow-up refactoring patch to change this magic number to a template parameter maybe

mdubet commented on Source/WebCore/platform/graphics/FontDescription.cpp:
```diff
@@ -32,6 +32,7 @@
 
 #include "FontCascadeDescription.h"
 #include "LocaleToScriptMapping.h"
+#include "platform/text/TextSpacing.h"
```
Do we really need this include ?

mdubet commented on Source/WebCore/platform/graphics/GlyphBuffer.h:
```diff
@@ -112,6 +112,15 @@ class GlyphBuffer {
         m_offsetsInString.append(offsetInString);
     }
 
+    void add(Glyph glyph, const Font& font, GlyphBufferAdvance advance, FloatPoint origin, GlyphBufferStringOffset offsetInString)
```
We should factorize this with the already existing add(...) function (default parameter for `origin`)

mdubet reviewed (approved):
The C++ part is ✅  Can't comment on the actual text implementation but I trust other informal reviewers (aka @fantasai)

vitorroriz committed:
text-autospace: implement ideograph-alpha for complex path within an text run
https://bugs.webkit.org/show_bug.cgi?id=277716
rdar://133309470

Reviewed by Matthieu Dubet.

This patch implements the processing of text-autospace: ideogram-alpha
only within an element. We don't yet handle element boundaries here.

Although we pass SpacingState context from one ComplexTextController
to another, we do that here in a limited way, only for measuring text for
layout and for painting. There are other places in code which this will
be necessary, for example, for handling element boundaries.

1. During the construction of ComplexTextController, we call ::adjustGlyphsAndAdvances
which already iterates through glyphs and adjust spacing for other reasons.
Now we process each pair of characters related to these glyphs here, adding the
spacing necessary before the "current" character.  For that reason, the SpacingState
stores information about the previous character of a run. We also save the measured
spacing in a new parallel vector m_textAutoSpaceSpacings. At this phase we can
only manipulate a glyph advance, however, for adding space "before" a glyph,
we need to move the glyph to the logical right, which is done later on ::advance.

2. ComplexTextController::advance is called for both layout and painting, but during
painting it has access to a GlyphBuffer and it add glyphs into it. We are introducing
a new GlyphBuffer::add function that also takes the glyph's origin, so we can manipulate
the origin as necessary by adding the previous calculated spacing.

3. Doing #1 and #2 is already enough for painting the extra spacing between relevant characters
according to their classes. Howeverm the width measured during layout would be broken because
IFC splits text content into inlineTextItem(s) and measure the width of each item independently.
This means that we already have to handle SpacingState passing here, otherwise we are not able
to handle spacing between characters on the boundary of different InlineTextItem.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-ideogram-alpha-001.html: Added.
* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isPunctuation):
(WTF::isOpeningPunctuation):
(WTF::isClosingPunctuation):
(WTF::isOfScriptType):
(WTF::eastAsianWidth):
(WTF::isEastAsianFullWidth):
(WTF::isCJKSymbolOrPunctuation):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::width):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.h:
(WebCore::Layout::TextUtil::width):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::ComplexTextController):
(WebCore::ComplexTextController::advance):
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/ComplexTextController.h:
(WebCore::ComplexTextController::ComplexTextRun::textAutospaceSize const):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::width const):
(WebCore::FontCascade::codePath const):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::add):
* Source/WebCore/platform/graphics/TextRun.cpp:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::WidthCache::add):
(WebCore::WidthCache::invalidateCacheForTextSpacing):
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
* Source/WebCore/platform/text/TextSpacing.cpp: Added.
(WebCore::TextAutospace::shouldApplySpacing const):
(WebCore::TextAutospace::textAutospaceSize):
(WebCore::TextSpacing::isIdeograph):
(WebCore::TextSpacing::isNonIdeographicNumeral):
(WebCore::TextSpacing::characterClass):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextAutospace::hasIdeographAlpha const):
(WebCore::TextAutospace::hasIdeographNumeric const):

Canonical link: https://commits.webkit.org/282192@main
